File: examples/cornell.cc
#include "nrtx/nrtx.h"

#include <cstdlib>
#include <iostream>
#include <string>

using namespace nrtx;

int main(int argc, char **argv) {
  const std::string out = argc > 1 ? argv[1] : "cornell.png";
  const int spp = argc > 2 ? std::atoi(argv[2]) : 256;
  const bool denoise = argc > 3 ? std::atoi(argv[3]) != 0 : true;

  Scene scene;
  const int white = scene.add_material({make_float3(0.73f, 0.73f, 0.73f), 0, 0.8f, 0, 1.5f, {}});
  const int red = scene.add_material({make_float3(0.65f, 0.05f, 0.05f), 0, 0.8f, 0, 1.5f, {}});
  const int green = scene.add_material({make_float3(0.12f, 0.45f, 0.15f), 0, 0.8f, 0, 1.5f, {}});
  const int light_mat = scene.add_material(
      {make_float3(0.0f, 0.0f, 0.0f), 0, 1.0f, 0, 1.5f, make_float3(15.0f, 15.0f, 12.0f)});
  const int glass = scene.add_material({make_float3(1.0f, 1.0f, 1.0f), 0, 0.0f, 1.0f, 1.5f, {}});
  const int metal = scene.add_material({make_float3(0.95f, 0.85f, 0.55f), 1.0f, 0.05f, 0, 1.5f, {}});

  // Room
  scene.add_mesh(make_quad(make_float3(0, 0, 0), make_float3(1, 0, 0), make_float3(0, 0, 1), white)); // floor
  scene.add_mesh(make_quad(make_float3(0, 1, 0), make_float3(1, 0, 0), make_float3(0, 0, 1), white)); // ceiling
  scene.add_mesh(make_quad(make_float3(0, 0, 1), make_float3(1, 0, 0), make_float3(0, 1, 0), white)); // back
  scene.add_mesh(make_quad(make_float3(0, 0, 0), make_float3(0, 0, 1), make_float3(0, 1, 0), red));   // left
  scene.add_mesh(make_quad(make_float3(1, 0, 0), make_float3(0, 0, 1), make_float3(0, 1, 0), green)); // right

  // Ceiling light (geometry + NEE)
  const float3 light_corner = make_float3(0.35f, 0.999f, 0.35f);
  const float3 light_u = make_float3(0.3f, 0, 0);
  const float3 light_v = make_float3(0, 0, 0.3f);
  scene.add_mesh(make_quad(light_corner, light_u, light_v, light_mat));
  scene.add_quad_light(light_corner, light_u, light_v, make_float3(15.0f, 15.0f, 12.0f));

  scene.add_mesh(make_box(make_float3(0.15f, 0.0f, 0.55f), make_float3(0.4f, 0.4f, 0.8f), white));
  scene.add_mesh(make_uv_sphere(make_float3(0.7f, 0.2f, 0.35f), 0.2f, glass));
  scene.add_mesh(make_uv_sphere(make_float3(0.35f, 0.15f, 0.3f), 0.15f, metal));

  scene.background_top = make_float3(0.0f, 0.0f, 0.0f);
  scene.background_bottom = make_float3(0.0f, 0.0f, 0.0f);

  Camera camera;
  camera.eye = make_float3(0.5f, 0.5f, -1.35f);
  camera.lookat = make_float3(0.5f, 0.5f, 0.5f);
  camera.fov_y_deg = 40.0f;
  camera.aspect = 1.0f;

  RenderConfig cfg;
  cfg.width = 800;
  cfg.height = 800;
  cfg.spp = spp;
  cfg.denoise = denoise;
  cfg.output_path = out;

  try {
    Renderer renderer;
    renderer.render(scene, camera, cfg);
  } catch (const std::exception &ex) {
    std::cerr << "Error: " << ex.what() << "\n";
    return 1;
  }
  return 0;
}

File: python/scenes/cornell.py
#!/usr/bin/env python3
"""Cornell Box scene for LumenCore."""
from __future__ import annotations

import sys

import lumencore as lc


def main() -> int:
    out = sys.argv[1] if len(sys.argv) > 1 else "cornell.avif"
    spp = int(sys.argv[2]) if len(sys.argv) > 2 else 256
    denoise = (int(sys.argv[3]) != 0) if len(sys.argv) > 3 else True

    scene = lc.Scene()
    white = scene.add_material(lc.Material(base_color=(0.73, 0.73, 0.73), roughness=0.8))
    red = scene.add_material(lc.Material(base_color=(0.65, 0.05, 0.05), roughness=0.8))
    green = scene.add_material(lc.Material(base_color=(0.12, 0.45, 0.15), roughness=0.8))
    light_mat = scene.add_material(
        lc.Material(base_color=(0, 0, 0), roughness=1.0, emission=(15, 15, 12))
    )
    glass = scene.add_material(lc.Material(base_color=(1, 1, 1), roughness=0.0, transmission=1.0, ior=1.5))
    metal = scene.add_material(lc.Material(base_color=(0.95, 0.85, 0.55), metallic=1.0, roughness=0.05))

    scene.add_mesh(lc.make_quad((0, 0, 0), (1, 0, 0), (0, 0, 1), white))
    scene.add_mesh(lc.make_quad((0, 1, 0), (1, 0, 0), (0, 0, 1), white))
    scene.add_mesh(lc.make_quad((0, 0, 1), (1, 0, 0), (0, 1, 0), white))
    scene.add_mesh(lc.make_quad((0, 0, 0), (0, 0, 1), (0, 1, 0), red))
    scene.add_mesh(lc.make_quad((1, 0, 0), (0, 0, 1), (0, 1, 0), green))

    light_corner = (0.35, 0.999, 0.35)
    light_u = (0.3, 0, 0)
    light_v = (0, 0, 0.3)
    # Emissive panel + QuadLight with MIS (visible lamp, no double-count).
    scene.add_mesh(lc.make_quad(light_corner, light_u, light_v, light_mat))
    scene.add_quad_light(light_corner, light_u, light_v, (15, 15, 12), use_mis=True)

    scene.add_mesh(lc.make_box((0.15, 0.0, 0.55), (0.4, 0.4, 0.8), white))
    scene.add_mesh(lc.make_uv_sphere((0.7, 0.2, 0.35), 0.2, glass))
    scene.add_mesh(lc.make_uv_sphere((0.35, 0.15, 0.3), 0.15, metal))

    scene.background_top = (0, 0, 0)
    scene.background_bottom = (0, 0, 0)

    camera = lc.Camera(eye=(0.5, 0.5, -1.35), lookat=(0.5, 0.5, 0.5), fov_y_deg=40, aspect=1.0)
    cfg = lc.RenderConfig(width=2048, height=2048, spp=spp, denoise=denoise, output_path=out)
    lc.Renderer().render(scene, camera, cfg)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
